[slider] Respect disabled property when already focused

avi-moskoff opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [X] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/next/CONTRIBUTING.md#sending-a-pull-request).

Hi, this isn't only my first PR to mui but also my first PR to any large open source project, so I'm happy to fill in any parts of the process that I've missed out on.

I have a reproduction of the problem this is supposed to fix on [this branch](https://github.com/pireads/material-ui/tree/disable-slider-repro). Running docs over there and changing the `Temperature` slider should disable it, but you can continue to change its value if you don't let go with your mouse or use arrow keys.

Thanks for making this project happen, it's been super clean to use so far :D

avi-moskoff committed:
Respect disable property when a key is pressed on a slider

avi-moskoff committed:
Also disable on handleTouchMove

eps1lon commented:
Could you add a test or codesandbox that illustrates the problem?

pireads commented:
@eps1lon Sure thing!

https://codesandbox.io/s/mui-pr-22247-eutvj?file=/src/App.js

eps1lon commented:
Makes sense. `<input type="range" />` works like expected.

I don't think we need the keydown handler since we blur the thumb if the slider becomes disabled.

Adding a unit test would still be nice though.

pireads commented:
@eps1lon It looks like the sandbox works as expected in the sandbox above for keydown on Chrome, but in both Firefox or Safari I am able to move the disabled slider with my arrow keys if I don't click on anything else after clicking the slider. I'll see if I can get a unit test going.

avi-moskoff committed:
Unit tests

avi-moskoff committed:
Allow an exception for chrome

Because chrome actually respects the blur, pressing a key on the thumb after disabling doesn't make sense.

eps1lon reviewed (dismissed):
We should add a test that solely checks if the slider gets blurred when it becomes disabled. That is sufficient to test if it can be operated with a keyboard.

eps1lon commented on packages/material-ui/src/Slider/Slider.test.js:
```diff
@@ -254,6 +254,68 @@ describe('<Slider />', () => {
       expect(getByRole('slider')).to.not.have.attribute('tabIndex');
     });
+
+    const SelfDisablingSlider = () => {
```
We try to keep test helpers local to the test since they tend to grow over time which makes debugging a specific test harder. Specifically here you don't need to re-use the component. The disabled-blur test can live with a standard `<Slider />` where you call `setProps({ disabled: true })` (returned from `render`).

eps1lon commented on packages/material-ui/src/Slider/Slider.test.js:
```diff
@@ -254,6 +254,68 @@ describe('<Slider />', () => {
+
+      const thumb = getByRole('slider');
+
+      expect(thumb).to.not.equal(document.activeElement);
```
We have a dedicated matcher which adds a more helpful message if the assertion fails
```suggestion
      expect(thumb).not.toHaveFocus();
```

avi-moskoff committed:
Changing the test was throwing an odd json error, but this works too

pireads commented:
Sorry, didn't see those comments before pushing, will modify the existing tests but I think if the CI fails here it will prove that the problem does exist on key press.

avi-moskoff committed:
PR comments, thanks Sebastian for your time!

oliviertassinari commented on packages/material-ui/src/Slider/Slider.js:
```diff
@@ -566,6 +570,10 @@ const Slider = React.forwardRef(function Slider(props, ref) {
   };
 
   const handleTouchMove = useEventCallback((nativeEvent) => {
+    if (disabled) {
+      return;
+    }
```
I don't think that this is a proper solution, the touch move will leak until the component gets unmounted. What about we add disabled in the list of dependencies for the event cleanup logic?

pireads replied:
Still pretty new to React, but I think I'm doing this now for the mouse. Not sure if you wanted it done for the keyboard as well/how to do it for the keyboard as well, but I can look into it if that's the direction you think it should go. Looking at the sandbox you posted above, it does look like firefox prevents the keyboard from making changes to the ranged input even without the blur.

oliviertassinari reviewed (commented):
Oh wow, right, Firefox nor Safari blur the active element if it gets disabled [(e.g. a textbox)](https://codesandbox.io/s/mui-pr-22247-forked-h151h?file=/src/App.js). I guess we shouldn't for this behavior either.

avi-moskoff committed:
cleanup tests

avi-moskoff committed:
Testing out cleanup logic

avi-moskoff committed:
Re-enable disable check on keydown

avi-moskoff committed:
lint

oliviertassinari committed:
reset active and focus visible states

oliviertassinari reviewed (approved):
I have added the same logic as https://github.com/mui-org/material-ui/blob/bf3fb8f5fb3576bb0a336f3bf5a4ae636286935b/packages/material-ui/src/ButtonBase/ButtonBase.js#L99-L101 
Happy to move forward, from my end

avi-moskoff committed:
Get rid of commented code

eps1lon commented on packages/material-ui/src/Slider/Slider.js:
```diff
@@ -430,9 +430,20 @@ const Slider = React.forwardRef(function Slider(props, ref) {
   const isRtl = theme.direction === 'rtl';
 
   const handleKeyDown = useEventCallback((event) => {
+    if (disabled) {
+      return;
+    }
```
This concern is still unadressed. If a disabled Slider is still focused then this should be fixed rather than ignoring keydown for these cases. If it's not focused anymore then we won't have to special case it here.

oliviertassinari replied:
A disabled Slider is still focused like a text box or a button would be: https://codesandbox.io/s/mui-pr-22247-forked-h151h?file=/src/App.js. Same reason why we have a disabled check in the key handler of the button base:
https://github.com/mui-org/material-ui/blob/0a84c3761b58019d99a02159c17ac5658d0cb585/packages/material-ui/src/ButtonBase/ButtonBase.js#L225

eps1lon replied:
Right this quirk again. Add a comment explaining that this is incorrect browser behavior. Disabled buttons should not be focused.

Then we can remove that branching in the test, or add appropriate expectations for each branch. One branch in the test is currently not covered by assertions.

oliviertassinari replied:
> Add a comment explaining that this is incorrect browser behavior. Disabled buttons should not be focused.

Are you saying it's a flaw in Firefox and Safari?

eps1lon replied:
> Elements that meet all the following criteria:
> [...]
> the element is not actually disabled;

-- https://html.spec.whatwg.org/multipage/interaction.html#focusable-area

Bug confirmed in https://bugzilla.mozilla.org/show_bug.cgi?id=1650092

avi-moskoff committed:
Add comments, more explicit testing

avi-moskoff committed:
Lint

eps1lon commented on packages/material-ui/src/Slider/Slider.js:
```diff
@@ -430,9 +430,23 @@ const Slider = React.forwardRef(function Slider(props, ref) {
   const handleKeyDown = useEventCallback((event) => {
+    if (disabled) {
+      // This check is necessary because Firefox and Safari will keep focus
+      // on a disabled slider, even though we blur it:
```
I think the new behavior is pretty confusing. We should either actually `blur()` it. Something like 

```js
useEnhancedEffect(() => {
  // doWeHaveARefToTheThumb?
  const { current: thumb } = whateverNameWeHaveGivenTheRef;
  if (disabled && thumb === document.activeElement) {
    thumb.blur()
  }
}, [disabled])
```

or not remove the focus-visible (and active?) stylings. 

Right now we remove the focus-visible style but in certain browsers it'll still be focused. That's confusing.

oliviertassinari replied:
The current behavior matches how our ButtonBase behaves and how a native button behaves (focus-ring gets removed, button stays focused). Unless there is references against this behavior in the web specifications, I think that we should respect the original design decisions of Firefox and Safari

eps1lon replied:
I understand the argument for consistency with ButtonBase. But this is incorrect behavior since it is not spec-compliant behavior and confirmed a bug in Firefox.

eps1lon replied:
> I think that we should respect the original design decisions of Firefox and Safari.

Then we shouldn't remove focus styling as browsers with these quirks don't do so as well.

oliviertassinari replied:
+1 for blurring https://github.com/mui-org/material-ui/pull/22247#discussion_r474223785

eps1lon replied:
Firefox does not remove the focus styling since the `:focus` pseudo class still applies. We just reset those styles and can't see it. A custom theme might still apply it.

oliviertassinari committed:
sebastian review

oliviertassinari committed:
Update packages/material-ui/src/Slider/Slider.test.js

Co-authored-by: Sebastian Silbermann <silbermann.sebastian@gmail.com>

eps1lon reviewed (dismissed):
> The closest I could find jsdom/jsdom#2931

Yeah, sorry about the confusion. A disabled element should no longer be the active element. But: It should also not receive a blur event when that happens. This behavior is covered by focus fixup rules:

> When the designated focused area of the document is removed from that Document in some way (e.g. it stops being a focusable area, it is removed from the DOM, it becomes expressly inert, etc.), designate the Document's viewport to be the new focused area of the document.

-- https://html.spec.whatwg.org/multipage/interaction.html#focus-fixup-rule

I don't find the specified behavior particularly helpful though I don't think we should dispatch them anyway. I tested it quickly in Firefox and they don't dispatch a blur event if a disabled, active  element is blurred: https://jsbin.com/makipogexi/2/edit?html,js,console,output

So everything is fine related in the blur-focus-keydown area. Just one question about touch remains

oliviertassinari committed:
skip specific tests

eps1lon committed:
link jsdom issue

eps1lon committed:
Add explicit test for a touch focusing the thumb

eps1lon committed:
Entangle effects

oliviertassinari commented on packages/material-ui/src/Slider/Slider.js:
```diff
@@ -679,6 +697,16 @@ const Slider = React.forwardRef(function Slider(props, ref) {
     };
   }, [handleTouchEnd, handleTouchMove, handleTouchStart]);
 
+  React.useEffect(() => {
+    if (disabled) {
+      const doc = ownerDocument(sliderRef.current);
+      doc.removeEventListener('mousemove', handleTouchMove);
+      doc.removeEventListener('mouseup', handleTouchEnd);
+      doc.removeEventListener('touchmove', handleTouchMove);
+      doc.removeEventListener('touchend', handleTouchEnd);
+    }
+  }, [disabled, handleTouchEnd, handleTouchMove]);
+
```
Not sure about the duplication of the event listener logic, it feels brittle to future changes.

What if we restored the previous approach but with a new branch like:

```jsx
if (disabled) {
  return undefined;
}
```

eps1lon replied:
Don't try to be clever with effects. The useEffect hook has cleanup logic built in for a reason. You're abusing cleanup methods for something that isn't added in the effect.

oliviertassinari replied:
I was thinking of going for:

```diff
diff --git a/packages/material-ui/src/Slider/Slider.js b/packages/material-ui/src/Slider/Slider.js
index 35a62d356b..73d0fa44b1 100644
--- a/packages/material-ui/src/Slider/Slider.js
+++ b/packages/material-ui/src/Slider/Slider.js
@@ -660,6 +660,10 @@ const Slider = React.forwardRef(function Slider(props, ref) {
   });

   React.useEffect(() => {
+    if (disabled) {
+      return undefined;
+    }
+
     const { current: slider } = sliderRef;
     slider.addEventListener('touchstart', handleTouchStart, {
       passive: doesSupportTouchActionNone(),
@@ -677,7 +681,7 @@ const Slider = React.forwardRef(function Slider(props, ref) {
       doc.removeEventListener('touchmove', handleTouchMove);
       doc.removeEventListener('touchend', handleTouchEnd);
     };
-  }, [handleTouchEnd, handleTouchMove, handleTouchStart]);
+  }, [disabled, handleTouchEnd, handleTouchMove, handleTouchStart]);

   const handleMouseDown = useEventCallback((event) => {
     if (onMouseDown) {
```

What do you feel as cleaver about it? disabled => no listeners

oliviertassinari commented on packages/material-ui/src/Slider/Slider.test.js:
```diff
@@ -160,7 +160,7 @@ describe('<Slider />', () => {
       const thumb = getByRole('slider');
       fireEvent.mouseDown(thumb);
-      expect(document.activeElement).to.equal(thumb);
+      expect(thumb).toHaveFocus();
```
Nice!

oliviertassinari commented on packages/material-ui/src/Slider/Slider.test.js:
```diff
@@ -254,6 +263,60 @@ describe('<Slider />', () => {
     });
+
+    it('should not respond to drag events after becoming disabled', function test() {
+      // TODO: Don't skip once a fix for https://github.com/jsdom/jsdom/issues/3029 is released.
```
Thanks for taking the time to open an issue

oliviertassinari committed:
avoid duplication
